File: samples/semgrep/Bad.cs
/*
 * INTENTIONALLY BAD CODE — DO NOT FIX.
 *
 * Layer 2 sample: bait for the semgrep/ ruleset. Deliberately kept OUT of the
 * samples/dotnet project so that adding Semgrep bait here never changes the
 * analyzer sample's expected error count.
 *
 * This file is never compiled — Semgrep only parses it.
 */

using System;
using System.Data;
using System.Data.SqlClient;
using System.Diagnostics;
using System.Security.Cryptography;
using System.Threading.Tasks;

namespace Maximalcode.SemgrepSample;

// --- no-float-for-money ----------------------------------------------------
public sealed class Invoice
{
    public double TotalAmount { get; set; }

    public float DiscountAmount { get; set; }

    // NOT flagged: decimal is the right type.
    public decimal NetTotal { get; set; }
}

public static class Billing
{
    // --- hardcoded-secret-dotnet -------------------------------------------
    private const string ApiToken = "ghp_A1b2C3d4E5f6G7h8I9j0";
    // A URL is exempt, but NOT when it carries userinfo — see bad.ts.
    private const string ConnectionString = "postgres://admin:hunter2is@db.internal:5432/prod";

    // Negative controls for #17 — these must stay SILENT.
    private const string TokenEndpoint = "https://oauth2.googleapis.com/token";
    private const string TokenNone = "none";

    // --- todo-without-issue ------------------------------------------------
    // TODO: switch this over to the new pricing service
    // TODO(#918): this one is fine — it has an issue and must NOT be flagged

    // --- no-ambient-clock (cross-language proof rule, C# side) -------------
    public static bool IsExpired(DateTime expiresAt) => expiresAt < DateTime.UtcNow;

    // --- no-float-for-money (parameter) ------------------------------------
    public static double ApplyTax(double amount) => amount * 1.19;

    // --- weak-crypto -------------------------------------------------------
    public static byte[] Fingerprint(byte[] body)
    {
        using MD5 hasher = MD5.Create();
        return hasher.ComputeHash(body);
    }

    // --- sql-string-concat-dotnet ------------------------------------------
    public static SqlCommand FindUser(string id)
    {
        return new SqlCommand($"SELECT * FROM Users WHERE Id = '{id}'");
    }

    // The CONCATENATION form of the same constructor. Interpolation had bait;
    // concatenation did not, and they are separate patterns.
    public static SqlCommand FindUserConcat(string id)
    {
        return new SqlCommand("SELECT * FROM Users WHERE Id = " + id);
    }

    // The Dapper / CommandText group: five entry points advertised by the rule,
    // and not one of them had a fixture. Both regex branches are exercised
    // across them — concatenation and interpolation are matched by different
    // expressions over raw source text, so covering one says nothing about the
    // other.
    public static void QueryUsers(IDbConnection conn, string id)
    {
        conn.Query("SELECT * FROM Users WHERE Id = " + id);
    }

    public static Task QueryUsersAsync(IDbConnection conn, string id)
    {
        return conn.QueryAsync($"SELECT * FROM Users WHERE Id = '{id}'");
    }

    public static void DeleteUser(IDbConnection conn, string id)
    {
        conn.Execute("DELETE FROM Users WHERE Id = " + id);
    }

    public static Task DeleteUserAsync(IDbConnection conn, string id)
    {
        return conn.ExecuteAsync($"DELETE FROM Users WHERE Id = '{id}'");
    }

    public static void SetCommandText(SqlCommand cmd, string id)
    {
        cmd.CommandText = "SELECT * FROM Users WHERE Id = " + id;
    }

    // NEGATIVE CONTROL for the Dapper group. A parameterised call must NOT
    // fire, or the rule is "any Query call" and gets switched off.
    public static void QueryUsersSafe(IDbConnection conn, string id)
    {
        conn.Query("SELECT * FROM Users WHERE Id = @id", new { id });
    }

    // --- command-injection-dotnet -----------------------------------------
    public static void Archive(string path)
    {
        Process.Start($"tar -czf backup.tgz {path}");
    }

    // --- sync-over-async ---------------------------------------------------
    public static string LoadBlocking(Task<string> work)
    {
        return work.GetAwaiter().GetResult();
    }

    // --- catch-and-swallow-dotnet -----------------------------------------
    public static void TryFlush(Action flush)
    {
        try
        {
            flush();
        }
        catch (Exception ex)
        {
        }
    }

    // The other two branches. `catch { }`, `catch (T e) { }` and `catch (T) { }`
    // are three separate AST shapes and the rule spells out all three — but
    // only the bound form had bait, and the bare form was "covered" solely by
    // the negative control below, which proves the escape hatch works and
    // nothing about the branch still matching.
    public static void TryFlushBare(Action flush)
    {
        try
        {
            flush();
        }
        catch
        {
        }
    }

    public static void TryFlushTyped(Action flush)
    {
        try
        {
            flush();
        }
        catch (InvalidOperationException)
        {
        }
    }

    // NEGATIVE CONTROL for catch-and-swallow-dotnet. This one must NOT fire.
    // See the matching note in bad.ts — the comment escape hatch the message
    // promises did not work until 2026-07-31. Kept in the BAD sample so the
    // existing finding assertion doubles as the regression test.
    public static void TryDispose(IDisposable resource)
    {
        try
        {
            resource.Dispose();
        }
        catch
        {
            // Disposal failures are not actionable here; the object is going away regardless.
        }
    }

    // The exception-filter form. C# lets a catch carry a `when (...)` clause
    // between the type and the block, and that clause is source text the
    // `pattern-not-regex` has to step over — until 2026-08-02 it did not, so a
    // FILTERED catch could not be cleared by the comment the message asks for
    // even though an unfiltered one could. The same bug as 2026-07-31, one
    // syntax further along, and found the same way: measuring against
    // Consumer A.
    //
    // Both halves are planted, per the lesson above the bare form: a negative
    // control on its own proves the escape hatch works and nothing about
    // whether the branch still matches. This one must FIRE.
    public static void TryFlushFiltered(Action flush)
    {
        try
        {
            flush();
        }
        catch (InvalidOperationException e) when (e.Message.Length > 0)
        {
        }
    }

    // NEGATIVE CONTROL for the filter form. This one must NOT fire.
    public static void TryDisposeFiltered(IDisposable resource)
    {
        try
        {
            resource.Dispose();
        }
        catch (InvalidOperationException e) when (e.Message.Length > 0)
        {
            // Disposal races on a shared handle are expected; the object is going away regardless.
        }
    }

    // --- debug-print-left-behind-dotnet -----------------------------------
    public static void Trace(string message)
    {
        Debug.WriteLine(message);
    }
}

File: samples/semgrep/bad.ts
/*
 * INTENTIONALLY BAD CODE — DO NOT FIX.
 *
 * Layer 2 sample: bait for the semgrep/ ruleset. Deliberately kept OUT of the
 * samples/typescript project so that adding Semgrep bait here never changes the
 * ESLint sample's expected finding count.
 *
 * This file is never compiled or linted — Semgrep only parses it.
 */

import * as childProcess from 'node:child_process';
import { exec, execSync } from 'node:child_process';
import crypto from 'node:crypto';

// --- todo-without-issue ----------------------------------------------------
// TODO: decide whether we keep this endpoint
// TODO(#412): this one is fine — it has an issue and must NOT be flagged

// --- hardcoded-secret-ts ---------------------------------------------------
const apiKey = 'sk_live_9f3b7c1d4e6a8b2c5d7e';
// A URL is exempt, but NOT when it carries userinfo — that is the shape that
// actually leaks, and the #17 value-guard must not blanket-exempt it.
const connectionString = 'postgres://admin:hunter2is@db.internal:5432/prod';

// Negative controls for #17 — these must stay SILENT. If either starts firing,
// the value-guard regressed back to name-only matching (0/5 precision).
const tokenEndpoint = 'https://github.com/login/oauth/access_token';
const UNASSIGNED_TOKEN = 'none';

// --- no-ambient-clock (cross-language proof rule, TS side) -----------------
export function isExpired(expiresAt: Date): boolean {
  return expiresAt < new Date();
}

// --- weak-crypto -----------------------------------------------------------
export function fingerprint(body: string): string {
  return crypto.createHash('md5').update(body).digest('hex');
}

// OpenSSL algorithm names are CASE-INSENSITIVE, so this is a working MD5 and
// was not matched by the literal-only rule. Pass 3.
export function signUpper(body: string): string {
  return crypto.createHash('MD5').update(body).digest('hex');
}

// The rule's message promises DES, but only the exact string "des" was listed.
// Triple-DES is des-ede3 / des-ede3-cbc, which is what real code writes.
export function encryptLegacy(key: Buffer, iv: Buffer) {
  return crypto.createCipheriv('des-ede3-cbc', key, iv);
}

// NEGATIVE CONTROLS for weak-crypto. Modern algorithms must NOT fire — a rule
// that flags every createHash call gets disabled, not fixed.
export function digestOk(body: string): string {
  return crypto.createHash('sha256').update(body).digest('hex');
}

export function encryptOk(key: Buffer, iv: Buffer) {
  return crypto.createCipheriv('aes-256-gcm', key, iv);
}

// --- sql-string-concat-ts --------------------------------------------------
export async function findUser(db: { query: (sql: string) => Promise<unknown> }, id: string) {
  return db.query(`SELECT * FROM users WHERE id = '${id}'`);
}

// The CONCATENATION branch, which had no fixture at all until security review
// pass 3 — so CI was green while `pattern-regex` matched only double quotes and
// the single-quoted form (the prevailing TypeScript style, and the commonest
// SQL-injection shape in JavaScript) went undetected. Both quote styles are
// baited now, because a regex over raw source text does not treat them alike.
export async function findUserDq(db: { query: (sql: string) => Promise<unknown> }, id: string) {
  return db.query("SELECT * FROM users WHERE id = " + id);
}

export async function findUserSq(db: { query: (sql: string) => Promise<unknown> }, id: string) {
  return db.query('SELECT * FROM users WHERE id = ' + id);
}

// The remaining two entry points from the accepted-method list. `.query` was
// the only one with a fixture, so `.execute` and `.raw` were advertised and
// untested — and the list itself is now a bound name plus a regex, which is
// the part that has to be exercised per name.
export async function findUserExecute(db: { execute: (sql: string) => Promise<unknown> }, id: string) {
  return db.execute("SELECT * FROM users WHERE id = " + id);
}

export async function findUserRaw(knex: { raw: (sql: string) => Promise<unknown> }, id: string) {
  return knex.raw(`SELECT * FROM users WHERE id = ${id}`);
}

// Prisma's raw escape hatches. The entry points used to be a fixed list of
// `.query` / `.execute` / `.raw`, so `$queryRawUnsafe` — the method whose NAME
// says it does not parameterise — matched nothing at all.
export async function findUserPrisma(
  prisma: { $queryRawUnsafe: (sql: string) => Promise<unknown> },
  id: string,
) {
  return prisma.$queryRawUnsafe('SELECT * FROM users WHERE id = ' + id);
}

export async function deleteUserPrisma(
  prisma: { $executeRawUnsafe: (sql: string) => Promise<unknown> },
  id: string,
) {
  return prisma.$executeRawUnsafe(`DELETE FROM users WHERE id = ${id}`);
}

// NEGATIVE CONTROL for sql-string-concat-ts. A parameterised query must NOT
// fire, or the rule is just "any query call" and will be switched off.
export async function findUserSafe(
  db: { query: (sql: string, params: unknown[]) => Promise<unknown> },
  id: string,
) {
  return db.query('SELECT * FROM users WHERE id = ?', [id]);
}

// NEGATIVE CONTROL for the widened entry point. `$queryRaw` is a TAGGED
// TEMPLATE and Prisma parameterises its interpolations, so it is the safe form
// of the same API and must stay out of the accepted-method regex.
export async function findUserPrismaSafe(
  prisma: { $queryRaw: (sql: TemplateStringsArray, ...values: unknown[]) => Promise<unknown> },
  id: string,
) {
  return prisma.$queryRaw`SELECT * FROM users WHERE id = ${id}`;
}

// --- command-injection-ts --------------------------------------------------
export function listDir(dir: string): void {
  exec(`ls -la ${dir}`, (err, stdout) => {
    // --- debug-print-left-behind-ts ---
    console.log(stdout);
  });
}

// The two CONCATENATION branches. Neither had a fixture, which is exactly why
// the single-quoted one was missing from the rule for as long as it was: a
// branch nothing bait tests cannot be observed to be absent.
export function listDirDq(dir: string): void {
  exec("ls -la " + dir, () => {});
}

export function listDirSq(dir: string): void {
  exec('ls -la ' + dir, () => {});
}

// ...and the other three entry points, none of which had one either.
export function listDirSync(dir: string): string {
  return execSync('ls -la ' + dir).toString();
}

export function listDirNs(dir: string): void {
  childProcess.exec(`ls -la ${dir}`, () => {});
}

export function listDirNsSync(dir: string): string {
  return childProcess.execSync("ls -la " + dir).toString();
}

// --- catch-and-swallow-ts --------------------------------------------------
export function loadSettings(raw: string): unknown {
  try {
    return JSON.parse(raw);
  } catch (e) {
  }
}

// The OPTIONAL-BINDING branch. `catch { }` and `catch (e) { }` are separate
// patterns because they are separate AST shapes, and only the bound form had
// bait — the bare form was covered solely by the negative control below, which
// proves the escape hatch works and nothing about the branch still matching.
export function loadSettingsBare(raw: string): unknown {
  try {
    return JSON.parse(raw);
  } catch {
  }
}

// NEGATIVE CONTROL for catch-and-swallow-ts. This one must NOT fire.
//
// The rule's message promises that a comment explaining the silence clears the
// finding. Until 2026-07-31 that promise was false — Semgrep matches the AST,
// comments are not AST nodes, so this block was still `{ }` to the pattern.
//
// It lives in the BAD sample deliberately: the suite asserts exactly what
// fires, so if the escape hatch ever breaks again this file gains a
// catch-and-swallow-ts finding it is not supposed to have and CI fails. No new
// assertion needed — the existing one IS the test.
export function tryClose(close: () => void): void {
  try {
    close();
  } catch {
    // Already closed — the caller cannot act on this and does not need to.
  }
}

// --- no-permission-denied-for-invisible-resource-ts -----------------------
class HttpError extends Error {
  constructor(
    public status: number,
    message: string,
  ) {
    super(message);
  }
}

export function getDocument(doc: { id: string } | undefined): { id: string } {
  if (!doc) {
    // Leaks that the document exists to anyone probing ids.
    throw new HttpError(403, 'Forbidden');
  }
  return doc;
}

// The other four branches. `!x`, `=== null` and `=== undefined` are three
// different AST shapes and the rule spells out all three; only the first had
// bait. The two express idioms below are the shapes this actually takes in a
// real handler, and they had none at all.
export function getDocumentNull(doc: { id: string } | null): { id: string } {
  if (doc === null) {
    throw new HttpError(403, 'Forbidden');
  }
  return doc;
}

export function getDocumentUndefined(doc: { id: string } | undefined): { id: string } {
  if (doc === undefined) {
    throw new HttpError(403, 'Forbidden');
  }
  return doc;
}

interface Res {
  status: (code: number) => Res;
  send: (body: string) => void;
}

export function sendDocument(doc: { id: string } | undefined, res: Res): void {
  if (!doc) {
    res.status(403);
    return;
  }
  res.status(200).send(doc.id);
}

export function returnDocument(doc: { id: string } | undefined, res: Res): void {
  if (!doc) {
    return res.status(403).send('Forbidden');
  }
  res.status(200).send(doc.id);
}

export { apiKey };
